From 6fc122cb73610c836242339d223b44d793a87ab8 Mon Sep 17 00:00:00 2001
From: Ran Shalit <rshalit@habana.ai>
Date: Tue, 31 Dec 2024 16:54:56 +0200
Subject: [PATCH] [FSW-8421] arch: fs1: use memcpy engine from common lib

No need for baremetal implementation of copy-engine,
as it is already implemented in common library.
Also, the common library implementation is more
complete as it also has a substitute as regular
memcpy instead this feature is not supported (as
for today it is not yet supported in coral)

Change-Id: I409abe3874924826d5539ce42d373dcb2280ec66
Signed-off-by: Ran Shalit <rshalit@habana.ai>
---

diff --git a/arch/fs1/arc/common.c b/arch/fs1/arc/common.c
index 26e233a..adab2c4 100644
--- a/arch/fs1/arc/common.c
+++ b/arch/fs1/arc/common.c
@@ -51,6 +51,7 @@
 #include <address_space.h>
 #include "common.h"
 #include "hl.h"
+#include <lib_common.h>
 
 static struct hl_pll_cfg pll_cfg[] = {
 	{
@@ -461,99 +462,9 @@
 #define MEM_CPY_HBW_SIZE_SHIFT_MASK	(BIT(MEM_CPY_HBW_SIZE_SHIFT) - 1)
 #define HL_MEM_CPY_TIMEOUT_US		10000
 
-int fs1_mem_copy(uint64_t src, uint64_t dst, uint32_t size, bool is_hbw, bool const_data)
-{
-	u64 start = timestamp_get_us();
-	u32 cpy_sts, copy_sz;
-	u32 cp_err, wr_err, rd_err;
-	u32 sz_limit, sz_shift, sz_mask;
-
-	if (is_hbw) {
-		sz_limit = MEM_CPY_HBW_SIZE_LIMIT;
-		sz_shift = MEM_CPY_HBW_SIZE_SHIFT;
-		sz_mask = MEM_CPY_HBW_SIZE_SHIFT_MASK;
-	} else {
-		sz_limit = MEM_CPY_LBW_SIZE_LIMIT;
-		sz_shift = MEM_CPY_LBW_SIZE_SHIFT;
-		sz_mask = MEM_CPY_LBW_SIZE_SHIFT_MASK;
-	}
-
-	if (size > sz_limit) {
-		preboot_puts("size exceeds MEM_CPY engine limit\n");
-		return -1;
-	}
-
-	regmap_writel((uint32_t)src & MEM_COPY_SRC_ADDR_LSB_BITS_31_4_M,
-		      mmIO_0_CFG_HALF_0_PSOC_PSOC_CFG_MEM_COPY_BASE + mmMEM_COPY_SRC_ADDR_LSB);
-	regmap_writel((uint32_t)(src >> 32), mmIO_0_CFG_HALF_0_PSOC_PSOC_CFG_MEM_COPY_BASE +
-					     mmMEM_COPY_SRC_ADDR_MSB);
-
-	regmap_writel((uint32_t)dst & MEM_COPY_DEST_ADDR_LSB_BITS_31_4_M,
-		      mmIO_0_CFG_HALF_0_PSOC_PSOC_CFG_MEM_COPY_BASE + mmMEM_COPY_DEST_ADDR_LSB);
-	regmap_writel((uint32_t)(dst >> 32), mmIO_0_CFG_HALF_0_PSOC_PSOC_CFG_MEM_COPY_BASE +
-					     mmMEM_COPY_DEST_ADDR_MSB);
-
-	/* set size of in granuality of 4 bytes */
-	copy_sz = size >> sz_shift;
-	if (size & sz_mask) {
-		preboot_puts("MEM_CPY size not aligned, rounding up\n");
-		copy_sz++;
-	}
-
-	regmap_writel(copy_sz, mmIO_0_CFG_HALF_0_PSOC_PSOC_CFG_MEM_COPY_BASE +
-			       mmMEM_COPY_COPY_SIZE);
-
-	/* use constant value write mode for memory init */
-	if (const_data)
-		regmap_writel(MEM_COPY_CONST_VAL_WR_MODE_VAL_M,
-			      mmIO_0_CFG_HALF_0_PSOC_PSOC_CFG_MEM_COPY_BASE +
-			      mmMEM_COPY_CONST_VAL_WR_MODE);
-	else
-		regmap_writel(0, mmIO_0_CFG_HALF_0_PSOC_PSOC_CFG_MEM_COPY_BASE +
-				 mmMEM_COPY_CONST_VAL_WR_MODE);
-
-	/* disable LBW size check per H15-255 */
-	regmap_rmwritel(mmIO_0_CFG_HALF_0_PSOC_PSOC_CFG_MEM_COPY_MSTR_IF_LBW_TRAN_CHK_BASE +
-			mmTRAN_CHK_LBW_CHECK_EN, 0, TRAN_CHK_LBW_CHECK_EN_SIZE_CHK_M);
-
-	/* Enable copy and wait for completion */
-	regmap_writel(MEM_COPY_COPY_EN_VAL_M, mmIO_0_CFG_HALF_0_PSOC_PSOC_CFG_MEM_COPY_BASE +
-					      mmMEM_COPY_COPY_EN);
-
-	/* let 'enable' command propagate before polling on status */
-	HL_MNG_W_BARRIER();
-
-	while (1) {
-		cpy_sts = regmap_readl(mmIO_0_CFG_HALF_0_PSOC_PSOC_CFG_MEM_COPY_BASE +
-				       mmMEM_COPY_COPY_EN);
-		/* COPY_EN is set by FW, and cleared by HW when copy ends */
-		if (!(cpy_sts & MEM_COPY_COPY_EN_VAL_M))
-			break;
-
-		if (timeout_expired(start, HL_MEM_CPY_TIMEOUT_US)) {
-			preboot_puts("error: MEM_CPY engine timeout\n");
-			return -1;
-		}
-	}
-
-	cp_err = regmap_readl(mmIO_0_CFG_HALF_0_PSOC_PSOC_CFG_MEM_COPY_BASE +
-			      mmMEM_COPY_COPY_ERR_STS);
-	wr_err = regmap_readl(mmIO_0_CFG_HALF_0_PSOC_PSOC_CFG_MEM_COPY_BASE +
-			      mmMEM_COPY_WR_ERR_STS);
-	rd_err = regmap_readl(mmIO_0_CFG_HALF_0_PSOC_PSOC_CFG_MEM_COPY_BASE +
-			      mmMEM_COPY_RD_ERR_STS);
-
-	if (cp_err | wr_err | rd_err) {
-		preboot_puts("MEM_CPY finished with errors\n");
-		return -1;
-	}
-
-	return 0;
-}
-
 int fs1_mem_init(uint64_t dst, uint32_t size, bool is_hbw)
 {
-	return fs1_mem_copy(0, dst, size, is_hbw, true);
+	return hw_mem_copy(0, dst, size, is_hbw, true);
 }
 
 u32 dev_spec_get_img_desc(struct hl_fw_img_desc *desc_ptr, enum fw_post_reset_cmd upd_status,
diff --git a/arch/fs1/arc/palad.c b/arch/fs1/arc/palad.c
index e70fa22..84f82ac 100644
--- a/arch/fs1/arc/palad.c
+++ b/arch/fs1/arc/palad.c
@@ -267,7 +267,7 @@
 	}
 
 	/* Copy SRAM_LOCATIONS to IO0_PCIE_FW_SRAM_FULL_BASE */
-	fs1_mem_copy(full_addr,
+	hw_mem_copy(full_addr,
 		IO0_PCIE_FW_SRAM_FULL_BASE,
 		(uint32_t)PHY_FW_SIZE, false, false);
 
@@ -356,7 +356,7 @@
 	 * exceeds MISC ICCM size (256kb), the rest is stored in MISC DCCM.
 	 * An efuse bit is controlling whether the application binary is larger than MISC ICCM size.
 	 */
-	fs1_mem_copy(IO0_MISC_ICCM_FULL_BASE, PREBOOT_TEXT_BASE_FULL, IO_MISC_ICCM_SIZE, false,
+	hw_mem_copy(IO0_MISC_ICCM_FULL_BASE, PREBOOT_TEXT_BASE_FULL, IO_MISC_ICCM_SIZE, false,
 		     false);
 
 	err = preboot_fuse_read(IO_EFUSE_PLDM_CFG_0_OFFSET, &val);
@@ -368,7 +368,7 @@
 	if (!val)
 		return;
 
-	fs1_mem_copy(IO0_MISC_DCCM_FULL_BASE, PREBOOT_TEXT_BASE_FULL + IO_MISC_ICCM_SIZE,
+	hw_mem_copy(IO0_MISC_DCCM_FULL_BASE, PREBOOT_TEXT_BASE_FULL + IO_MISC_ICCM_SIZE,
 		     IO_MISC_DCCM_SIZE, false, false);
 }
 
@@ -393,7 +393,7 @@
 	 * Since SPSRAM content is erazed on cold reboot, fresh copy stored in MISC DCCM and copied
 	 * back to the SPSRAM as part of the preparation to run preboot rtos.
 	 */
-	fs1_mem_copy(IO0_MISC_DCCM_FULL_BASE, IO0_SP_SRAM_FULL_BASE, BOOT_INFO_SIZE, false, false);
+	hw_mem_copy(IO0_MISC_DCCM_FULL_BASE, IO0_SP_SRAM_FULL_BASE, BOOT_INFO_SIZE, false, false);
 }
 
 void btldr_copy_data(void)
